### PR TITLE
Add typed batch Ledger Entry payloads

### PR DESCRIPTION
**Pull request by vigneshwerv**
Implements the shared [typed-batch-entries spec](https://github.com/fragment-dev/graphql-queries/pull/15) (v0.1.0). Codegen now derives a typed payload struct per `(entry type, typeVersion)` from the `addLedgerEntry` operations the Fragment CLI generates, so a batch is built with compile-time parameter checking instead of hand-built untyped maps.

Stacked on #30, which supplies the `AddLedgerEntries` operation.

## What's here

- `internal/typedentries/` — derivation (§2.1–2.5) and Go emission
- `batch/` — `Entry` plus an ordered JSON writer that omits unset fields. `LedgerEntryInput` can't carry a conforming batch: it has no `omitempty`, so unset fields go out as explicit `null`. `omitempty` isn't the fix either — §3.2 rules it out for conflating `""`/`0`/`false` with unset.
- `queries.AddTypedLedgerEntries` and `queries.RawEntry` for §3.5 mixing
- Payloads land in a nested `typed_payloads` package. That's what makes §2.6's keyed-literal requirement enforceable: Go only rejects an unkeyed literal assigning to an unexported field from *outside* the declaring package, so payloads in the caller's own package would have no protection.
- `docs/spec-conformance.md` — section-by-section mapping and deviations

## Deviations

- **No conformance runner (§5).** Go can't do it in one process — payloads are generated source and must be compiled before anything can construct them. The fixtures are covered by tests over committed generated output, which doubles as the required snapshot test. Cost: this repo won't go red on its own when upstream changes a fixture; re-syncing `internal/conformance/testdata/` is manual for now.
- **Optionals are pointers (§3.2).** nil means unset. Explicit `null` isn't expressible on a typed payload; `RawEntry` is the escape hatch, which §3.5 already blesses.
- **Baseline profile, not strict (§3.4).** Strict needs a replacement `graphql.Client` with `SetEscapeHTML(false)`. Under baseline the comparison is parsed-JSON, so escaping isn't observable and `006-non-ascii` passes without it.

## Verification

`go test ./...`, `go vet`, `gofmt` clean. All six fixtures pass. `make codegen` still produces no `typed_payloads/` for the SDK's own operations (its `addLedgerEntry` binds `type: $entryType`), so `queries.go` is byte-identical.

Also compiled and ran the generated code from a real downstream module — worth doing, because two bugs here only surface at compile time and were masked by tests supplying values the CLI doesn't:

1. `String!` parameters generated `queries.String`, which doesn't exist — built-in GraphQL scalars aren't in the generator's binding table. Fixed with a fallback; `String`/`Int` removed from the test scalar maps so they exercise the production path.
2. A `JSON`-scalar parameter emitted `json.RawMessage` without importing `encoding/json`. `Emit`'s gofmt pass doesn't catch a missing import.

## Verified against a live API

`internal/livetest` stores a Schema, creates a Ledger, and posts for real. Runs in CI as the `Live API tests` job. This answers spec §6's known gap: **an entry object with `type` present and `lines` absent is accepted.** Atomicity and per-entry ik replay behave as specified.

It also found a bug no offline test could: raw entries kept their `null`s, so a raw `LedgerMatchInput` encoded as `{"id":null,"ik":"L"}` while a typed payload sent `{"ik":"L"}` — and the API resolves those as *different* Ledgers, rejecting any batch that mixed the two. `RawEntry` now omits unset fields. In Go that costs nothing, since a nil `*string` in the generated input types means both unset and null, so there was never an explicit null to preserve.

## For graphql-queries#15

Three suggestions filed, all recorded in the deviation register:

1. **§2.1** — an operation with a literal `type` but a variable `typeVersion` has no defined identity. §2.1 doesn't require the version to be a literal, but §2.2 derives identity from "the integer literal … if present, otherwise 1". Treating it as absent would pin to v1 though the operation accepts any version, so this SDK skips it with a warning. Suggested a seventh recognition condition.
2. **§3.5** — the accepted null asymmetry makes the mixing MUST unsatisfiable, per the live finding above. Suggested that raw inputs must omit unset fields too. **This one determines whether the `RawEntry` behaviour here is conformant or a deviation.**
3. **§5** — cites `spec/conformance/`, while `README.md` vendors to `shared-spec/` with the runner as a sibling `spec/`. Independently raised by another reviewer.

Also fed back that our live run confirms `typeVersion: 1` is accepted, which was an open question on that PR, and documented the 30-Ledger-Line batch cap a reviewer found in the API implementation — the SDK cannot enforce it, since line count per entry type never reaches the generated operations.

### CODE REVIEW

File: batch/batch.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// Entry is a single element of a batch. It is implemented by the generated
+// typed payloads and by
+// [github.com/fragment-dev/fragment-go/v4/queries.RawEntry].
+//
+// FragmentBatchEntry is a marker, so that an arbitrary [encoding/json.Marshaler]
+// cannot be passed to a batch by accident. It is exported only because generated
+// payloads live outside this package and so cannot implement an unexported
+// method.
+type Entry interface {
+	json.Marshaler
```
Comment (vigneshwerv) [inline batch/batch.go:27]: Basically means it has a `MarshalJSON()` method implemented for the type.

File: batch/batch.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	FragmentBatchEntry()
```
Comment (vigneshwerv) [inline batch/batch.go:28]: This is branding the types.

File: internal/typedentries/derive_test.go [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+	if payloads[0].GoName != "TV1Entry" {
+		t.Errorf("GoName = %q, want TV1Entry", payloads[0].GoName)
+	}
+}
+
+// TestUnpinnedVersionNormalisesToOne covers the requirement that an operation
+// pinning no typeVersion is treated as version 1 everywhere: identity, name and
+// wire value. An entry with no typeVersion resolves to 1 server-side rather than
+// to the latest version, so a payload named V1 that posted no version would
+// mislead anyone reading it.
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go]: think the last sentence of this is kind of confusing. is it saying if the entry DIDN'T resolve no typeVersion -> V1, that we'd get a payload named V1 (bc of the server side resolution) but it wouldn't have a version anywhere else (where?) and that's misleading?

File: internal/typedentries/derive_test.go [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+		  addLedgerEntry(ik: $ik, entry: {ledger: {ik: $ledgerIk}, type: "t", typeVersion: 1, parameters: {amount: $amount}}) { __typename }
+		}
+		mutation B($ik: SafeString!, $ledgerIk: SafeString!, $amount: String!, $fee: String!) {
+		  addLedgerEntry(ik: $ik, entry: {ledger: {ik: $ledgerIk}, type: "t", typeVersion: 2, parameters: {amount: $amount, fee: $fee}}) { __typename }
+		}`)
+
+	if len(payloads) != 2 {
+		t.Fatalf("expected 2 payloads, got %d", len(payloads))
+	}
+	if payloads[0].GoName != "TV1Entry" || payloads[1].GoName != "TV2Entry" {
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go]: can we actually check if TV1Entry and TV2Entry actually got the right parameters? maybe it's not necessary though

File: internal/typedentries/derive_test.go [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+// payload is still emitted when the parameters cannot be typed individually.
+func TestParametersWithoutInlineObjectFallBackToUntyped(t *testing.T) {
+	for name, src := range map[string]string{
+		"parameters bound to a variable": `
+			mutation A($ik: SafeString!, $ledgerIk: SafeString!, $parameters: JSON!) {
+			  addLedgerEntry(ik: $ik, entry: {ledger: {ik: $ledgerIk}, type: "t", parameters: $parameters}) { __typename }
+			}`,
+		"no parameters at all": `
+			mutation A($ik: SafeString!, $ledgerIk: SafeString!) {
+			  addLedgerEntry(ik: $ik, entry: {ledger: {ik: $ledgerIk}, type: "t"}) { __typename }
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go:305]: why would the no parameters at all case give us untyped params in the payload? I would expected them not to be there at all, similar to how the fixed params are skipped.

File: internal/typedentries/derive_test.go [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+// TestParameterTypesComeFromVariableDefinitions covers the requirement that a
+// parameter's type and required-ness are read from the variable definition rather
+// than guessed from the field name.
+func TestParameterTypesComeFromVariableDefinitions(t *testing.T) {
+	payloads, _ := derive(t, `
+		mutation A(
+		  $ik: SafeString!
+		  $ledgerIk: SafeString!
+		  $required: String!
+		  $optional: String
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go:349]: if this is testing that it's not just guessing from the name, should $required be optional and $optional be required? like the way it is now, if $required is made required in the payload, how do we know if it's doing it correctly or guessing

File: internal/typedentries/derive_test.go [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+	if err == nil {
+		t.Fatal("expected an error for two entry types producing one identifier")
+	}
+	if !strings.Contains(err.Error(), "same Go identifier") {
+		t.Errorf("unexpected error: %v", err)
+	}
+}
+
+// TestCommonFieldsAreFixed pins the requirement that the common-field set comes
+// from LedgerEntryInput rather than from the source operation.
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go]: source being mutation gql doc that was produced from the user's gql schema? like no matter what the entry is, we always get the common fields (Ik, ledgerIk, etc) from a fixed list, rather than from the entry that's being passed in?

File: internal/typedentries/derive_test.go [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+// the operation would still produce the right JSON for any fixture whose operation
+// happens to bind those fields. The guarantee is about where the list comes from,
+// so the list itself is what has to be asserted.
+func TestCommonFieldsAreFixed(t *testing.T) {
+	var got []string
+	for _, f := range CommonFields {
+		got = append(got, f.Name)
+	}
+
+	want := []string{"Ik", "LedgerIk", "Posted", "Description", "Tags", "Groups", "Conditions"}
```
Comment (sophieleewu) [inline internal/typedentries/derive_test.go]: confused what this is testing. looks like `got` is just an arr of the common fields and then we're comparing it to `want` which is also just a list of the common fields??

File: internal/typedentries/snapshot_test.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	return source
+}
+
+// TestGeneratedPayloadsAreCurrent is the snapshot test: generated identifiers and
+// signatures cannot change without a diff here.
+//
+// That matters because callers write those identifiers by hand. An accidental
+// rename is a breaking change for every call site, and nothing else in the test
+// suite would notice.
+func TestGeneratedPayloadsAreCurrent(t *testing.T) {
```
Comment (sophieleewu) [inline internal/typedentries/snapshot_test.go:88]: why is this file called golden_test? maybe snapshot_test would be more explicit

File: internal/typedentries/emit.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+	comment(b, "", fmt.Sprintf("FragmentBatchEntry marks %s as usable in a batch.", p.GoName))
+	fmt.Fprintf(b, "func (%s) FragmentBatchEntry() {}\n\n", p.GoName)
+
+	emitMarshal(b, p)
+}
+
+// paramDoc describes a parameter, always naming the value it travels under. That
+// is the fact a reader most needs and cannot recover from the field name, since
+// even the routine case renames user_id to UserId.
```
Comment (sophieleewu) [inline internal/typedentries/emit.go]: huh? not sure what this is getting at. is it just renaming/validating params?

maybe something like 
`// paramDoc formats an entry's parameters, handling naming collisions blah blah blah...`

File: queries/batch_test.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+	for _, want := range []string{`"ik":""`, `"type":""`, `"typeVersion":0`, `"tags":[]`} {
+		if !strings.Contains(string(got), want) {
+			t.Errorf("expected %s to survive, got: %s", want, got)
+		}
+	}
+}
+
+// TestRawEntryWithLines covers the case RawEntry exists for: an entry with lines
+// instead of a type, which no typed payload can express.
```
Comment (sophieleewu) [inline queries/batch_test.go:123]: if this is the only test that covers entry with lines I wonder if we want to add another case or two or expand it a little. like does it preserve multiple lines instead of just one for example

File: internal/generated/keyed_literal_test.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+			fields:      `struct{}{}, "ik-1", "prod", nil, nil, nil, nil, nil, "u", "o", "USD", "1"`,
+			wantCompile: false,
+		},
+		{
+			name: "partial unkeyed literal is rejected",
+			// Go rejects an unkeyed literal that omits fields on its own, but
+			// this also confirms the caller cannot get partway there.
+			fields:      `struct{}{}, "ik-1", "prod"`,
+			wantCompile: false,
+		},
```
Comment (sophieleewu) [inline internal/generated/keyed_literal_test.go]: what about this test case is "partial"? like it omits other fields like `LedgerIk`?
I would think a partial unkeyed literal would look like

```
fields:      `struct{}{}, "Ik: ik-1", "prod"`,
```

File: internal/generated/wire_test.go [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+		//    share a batch with the typed payloads above: the API resolves
+		//    {"id":null,"ik":"prod"} as a different Ledger from {"ik":"prod"}.
+		queries.RawEntry{Input: queries.AddLedgerEntryInput{
+			Ik: "ik-5",
+			Entry: queries.LedgerEntryInput{
+				Ledger:     &queries.LedgerMatchInput{Ik: ptr("prod")},
+				Type:       &entryType,
+				Parameters: &rawParameters,
+			},
+		}},
```
Comment (sophieleewu) [inline internal/generated/wire_test.go:164]: if the point of the `RawEntry` is to let you post lines, should we have a test case that has lines on a raw entry?

File: internal/livetest/add_ledger_entries_test.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+		"driver_id":     "driver-2",
+	})
+	entryType := "order_placed"
+	raw := queries.RawEntry{Input: queries.AddLedgerEntryInput{
+		Ik: "mixed-raw-" + rawOrderID,
+		Entry: queries.LedgerEntryInput{
+			Ledger:     &queries.LedgerMatchInput{Ik: &ledgerIk},
+			Type:       &entryType,
+			Parameters: &rawParams,
+		},
```
Comment (sophieleewu) [inline internal/livetest/add_ledger_entries_test.go:167]: would it be good to test this works when posting lines too?

File: internal/livetest/add_ledger_entries_test.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+		}
+	}
+}
+
+// TestAddTypedLedgerEntriesReportsIkReplay reposts a batch unchanged.
+//
+// Idempotency keys are per entry rather than per batch, so a resend must commit
+// nothing new and report every entry as a replay. This is what makes a retry after
+// an ambiguous failure safe.
+func TestAddTypedLedgerEntriesReportsIkReplay(t *testing.T) {
```
Comment (sophieleewu) [inline internal/livetest/add_ledger_entries_test.go:92]: should we test if one in a batch is a replay but the other isn't?

File: internal/livetest/test_helpers.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
+package livetest
```
Comment (sophieleewu) [inline internal/livetest/test_helpers.go:22]: this file is named like other test files (`_test` suffix) but seems to just be helpers for the tests themselves.
maybe rename to `test_helpers.go`

**Review by sophieleewu — APPROVED** [review]

File: internal/typedentries/derive_test.go (hunk 5 above)
Comment (vigneshwerv) [reply internal/typedentries/derive_test.go:305]: Great catch! I think we'd want to generate a typed version without any parameters. If `$lines` is not included by the CLI, then this is *not* a runtime entry; just an entry with no parameters.

File: internal/typedentries/derive_test.go (hunk 6 above)
Comment (vigneshwerv) [reply internal/typedentries/derive_test.go:349]: `$required` is of type `String!` and not `String`, so it's required. The `!` in GraphQL denotes a required value.

File: internal/typedentries/derive_test.go (hunk 7 above)
Comment (vigneshwerv) [reply internal/typedentries/derive_test.go]: That's right. It probably makes sense to derive these from the operation, but this is consistent with the other SDKs at the moment.

File: internal/livetest/add_ledger_entries_test.go (hunk 15 above)
Comment (vigneshwerv) [reply internal/livetest/add_ledger_entries_test.go:92]: I think the point here is to test that `IkReplay` is parsed correctly in the SDK. In-depth testing of the API _behavior_ should live closer to the API, imo. What'd you think?

File: internal/typedentries/derive_test.go (hunk 8 above)
Comment (vigneshwerv) [reply internal/typedentries/derive_test.go]: Yeah, good catch! I think the point of this test is to ensure that a user can set all of the attributes on `LedgerEntryInput`. Updated the test to use reflection and compare fields, instead of comparing two hard-coded lists.

File: internal/generated/keyed_literal_test.go (hunk 12 above)
Comment (vigneshwerv) [reply internal/generated/keyed_literal_test.go]: Oh, partial 'cause we haven't provided all the parameters!
